### ISSUE TITLE
Paper Mario crash when loading save states

### ISSUE THREAD

**Issue by fzurita** [mupen64plus/mupen64plus-core#197]
Paper Mario sometimes crashes when loading save states. I suspect that it's because save states don't contain the TLB state so it's not being re-enabled quickly enough. This crashes can happen when loading a save state and then pushing that start button.

This was a game crash, not a core crash, in game screenshot:
![screenshot_2016-10-16-13-41-23](https://cloud.githubusercontent.com/assets/11581687/19419495/cf718c60-93a6-11e6-8738-b2e219884191.jpg)

**Comment by Gillou68310** [comment]
I forced using_tlb to 0 in the code but couldn't reproduce the issue. Do you have the savestate when the problem occurs?
 
BTW once set the using_tlb flag is only cleared on core shutdown, is this happening when using the resume option in mupen64plus-ae?

**Comment by fzurita** [comment]
Yes, this is happening during the resume option. I can't seem to repeat it consistently, only sometimes. I'll see if I can sing a way to make it happen more often.

**Comment by Gillou68310** [comment]
Damn I hate this resume option, we need to find a way to make it more predictable, for example only do a loadstate on first frame. This is what the --savestate option is doing btw.

**Comment by Gillou68310** [comment]
See mupen64plus-ui-console for the --savestate option

**Comment by fzurita** [comment]
Ok

**Comment by fzurita** [comment]
Why are we using ```M64CMD_STATE_LOAD``` instead of ```--savestate``` for resume? Any idea? It would be very easy to change to ```--savestate``` instead.

Also, is the "slot" functionality doing it correctly?

**Comment by Gillou68310** [comment]
The drawback of the ```--savestate``` option is that the first frame will be shown before the loadstate happens. I think @littleguy77 wanted to avoid this.

> Also, is the "slot" functionality doing it correctly?

What do you mean?

**Comment by fzurita** [comment]
You know how we have the option to save and load states using slots? I mean saving and loading using that functionality.

**Comment by Gillou68310** [comment]
You can't load slot using the ```--savestate``` option. You need to specify the path to the savestate ```--savestate C:\save.state```

**Comment by fzurita** [comment]
Ok

**Comment by Gillou68310** [comment]
Otherwise you can set your own frame callback ```coreDoCommand(M64CMD_SET_FRAME_CALLBACK, 0, (void*)MyFrameCallback)``` and do whatever you want inside it. You'll have to wait before M64CORE_EMU_STATE changed to M64EMU_RUNNING using OnStateCallbackListener before setting the callback.

**Comment by fzurita** [comment]
OK, if you think that would get around this bug.

**Comment by Gillou68310** [comment]
I'm not saying that this will fix the issue. However with the ```--savestate``` option, the loadstate will always happen at the same moment, so if you manage to reproduce the issue once, the crash should happen every time you try to resume the corresponding savestate.

**Comment by Gillou68310** [comment]
It's just that I won't be able to fix the issue if it's not 100% reproducible

**Comment by Gillou68310** [comment]
The implementation is pretty simple:
if( !sIsRestarting)
{
    arglist.add( "--savestate" );
    arglist.add( saveToLoad );
}
and remove the whole block after the ```// Auto-load state if desired``` comment.
We still have to find a way to avoid showing the first frame before the loadstate happens.
You can test with dk64, the nintendo logo will always be shown before the loadstate happens.

**Comment by Gillou68310** [comment]
@richard42 I'd like to change the ```--savestate``` option implementation. The idea is to load the savestate on the first safe state interrupt instead of loading it on the first frame. Do you think this is an acceptable solution?

Code here: 
https://github.com/Gillou68310/mupen64plus-core/commit/46654f26c927d2863d58680708ac7ed5c704f1e6
https://github.com/Gillou68310/mupen64plus-ui-console/commit/334db80b5559709905bd150b5f20fbf4c4005f49

**Comment by Narann** [comment]
I added the `--savestate` at startup option for my regression test framework (yet to be published):

https://github.com/mupen64plus/mupen64plus-ui-console/commit/566c9f9903592a1bda9a1e3f299537cb7027d9ac
https://github.com/mupen64plus/mupen64plus-ui-console/commit/7d5247b3990a5b8232fb717af43355d10a2a3561
https://github.com/mupen64plus/mupen64plus-ui-console/commit/c495ff81e68a45b7c703629c20b88a5a56031a9f

The point is to work like this:

* Load the state
* Wait a given number of frame
* Do a screenshot
* Quit

The command is generated by a python script like this:
```python
cmd = ["./mupen64plus",
   "--resolution", "640x480",
   "--audio", "dummy",
   "--configdir", os.path.join(current_dir, "config"),
   "--nosaveoptions",
   "--testshots", ",".join((str(f) for f in frames)),
   "--savestate", os.path.join(save_state_dir, "%s.st0" % test_name),
   "--sshotdir", tmp_screenshot_dir,
   rom_path]
```
I saw your code and it should work fine. I will try it asap and keep you in touch. :)

**Comment by Narann** [comment]
I just tested. For an unknown reason, I have a black screen or no texture using `--nosaveoptions` with rice. mupen64plus-video-glide64mk2 doesn't seems to have any problem so I guess there is a problem with Rice options.

Anyway, It's ok for me. I suggest to do PR. :)

**Comment by Gillou68310** [comment]
Done ;-)
What is the purpose of the ```--nosaveoptions``` in rice?

**Comment by Narann** [comment]
It's supposed to be a core option to do not save mupen64plus.cfg. I don't really understand why it create some troubles with rice.

**Comment by Narann** [comment]
Can you confirm latest PR fixed you paper mario bug @fzurita? :)

**Comment by fzurita** [comment]
Sure, I'll check it out. This was not 100% reproducible, so it will be hard to 100% verify. But I will keep playing through Paper Mario and see if it happens again.

**Comment by fzurita** [comment]
Ok, I just had the in-game crash again after including the change to ui-console and the core. When I tried to manually load the saves state that caused it, it didn't crash... It's frustrating to reproduce the issue since it's so inconsistent.

**Comment by fzurita** [comment]
Whoops, forgot that part.

**Comment by Gillou68310** [comment]
No problem ;-)

**Comment by fzurita** [comment]
Ok, I haven't had it happen when resuming yet. But I just had it happen during a slot save. Here's is there slot save, just load it and push the N64 start button.

[Paper Mario (U) [!].st0.zip](https://github.com/mupen64plus/mupen64plus-core/files/546296/Paper.Mario.U.st0.zip)

**Comment by fzurita** [comment]
Ok, upon restarting the app and loading the save state. The game no longer crashes.

Also I couldn't get the app to stop crashing until I deleted all game data. I should have captured the core dump.

**Comment by fzurita** [comment]
Ok, I just had it happen after using the --savestate option. Once this in-game crash happens, it doesn't matter what states I load. Only way to recover is to kill then restart the app and I'm able to load the save state just fine and get into the in-game subscreen.

This crash only has happened after entering the in-game subscreen.

**Comment by Gillou68310** [comment]
Ok I confirm that this bug is caused by the ```using_tlb``` flag. When loading the first savestate you provided using the ```--savestate``` option I'm getting the issue 100% of the time.

![sans titre](https://cloud.githubusercontent.com/assets/2852068/19640736/607381b8-99de-11e6-8323-83a62f231dee.png)

Initializing the ```using_tlb``` flag to 1 fix the issue.
So like you said, the only way to fix this issue is to add the ```using_tlb``` flag to the savestate.

@Narann, @richard42 is the following solution ok for you?
https://github.com/Gillou68310/mupen64plus-core/commit/22aa4f41b2c0354b260824a43d433ae3de81266f

**Comment by fzurita** [comment]
I saw your solution, what would happen if we load an old format save state? I don't see a check to make sure the save state is of the correct size. I would imagine that if we loaded an old save state that doesn't have the using TLB field, then it should still work?

**Comment by Gillou68310** [comment]
When an old savestate is loaded ```gzread(f, newData, sizeof(newData));``` will do nothing because the end of the file has already been reached at this time (```f->have == 0```). Since ```newData``` is initialized with zeros ```using_tlb``` will be initialized to 0 when the following code will be called:

```
curr = newData;
using_tlb = GETDATA(curr, unsigned int);
```

**Comment by fzurita** [comment]
Ok, gotcha, I wasn't aware that gzread worked that way.

What would be more interesting is what's going to happen if more fields are added that are not specific to the new_dynarec. Maybe the space should be reserved even if the new_dynarec is not being used.

Another option could be that the TLB flag could be a game specific configuration. Didn't an issue like that get created?

**Comment by Gillou68310** [comment]
> What would be more interesting is what's going to happen if more fields are added that are not specific to the new_dynarec. Maybe the space should be reserved even if the new_dynarec is not being used.

If new fields are added, they will probably not be added to the end of the file like I did, most likely a new savestate format will be created.

> Another option could be that the TLB flag could be a game specific configuration. Didn't an issue like that get created?

TLB isn't always activated at the beginning of a game, it could happen later in the game. Forcing TLB when it's not needed may break things.

**Comment by richard42** [comment]
the savestate header has a version number in it.  I added this when we changed the format, so that we could handle backwards compatibility.  This version number is currently 0x00010000. I would suggest bumping it up by a minor number (say, to 0x00010100) and then adding the extra flag at the end.  You can tell be the version number when loading the state if the new flag is expected to be there or not.

**Comment by Narann** [comment]
I have hundreds of MB of  savestates I spend time to generate for my regression test suite. I would love to keep backward compatibility. And from what @Gillou68310 said, it doesn't seems to be complicated.

**Comment by Gillou68310** [comment]
My implementation is already backward compatible, bumping the version number will break that  backward compatibility.

```
if(version != 0x00010000)
{
    main_message(M64MSG_STATUS, OSD_BOTTOM_LEFT, "State version (%08x) isn't compatible. Please update Mupen64Plus.", version);
    gzclose(f);
    SDL_UnlockMutex(savestates_lock);
    return 0;
}
```

**Comment by richard42** [comment]
The proper way to do it is to increment the version number and then handle both cases explicitly.  Don't set a bad precedent by changing the format but leaving the version number the same.

If you want to support forward compatibility (older programs working with newer savestates), then you could add this feature (to new builds) by modifying this version check to mask out the upper 16 bits and only comparing them.  Then, in the future, if we make a change which would break forward compatibility, then you bump up the top 16 bits, otherwise bump up the lower 16 bits.  This is what we do for the plugin versioning.

**Comment by Gillou68310** [comment]
I created the PR following your indications ;-)
https://github.com/mupen64plus/mupen64plus-core/pull/199

**Comment by fzurita** [comment]
I haven't seen a crash for a few days now. So I think it's fixed. I'll reopen if I see it happen again.

**Comment by fzurita** [comment]
This is still happening but much less often.

**Comment by Gillou68310** [comment]
The TLB flag should never be turned off after being turned on. Could you provide a save state?

**Comment by fzurita** [comment]
Next time it happens I'll capture the save state.

**Comment by Gillou68310** [comment]
Thanks ;-)

**Comment by Gillou68310** [comment]
Can't reproduce it on windows :-(
Also the TLB flag is set to 1 in the savestate you provided. Forcing it to 0 doesn't trigger the issue BTW.

Are you sure you shared the right savestate? :-p

**Comment by fzurita** [comment]
Yeah, very sure. Also, I was able to load that same save state later and I didn't have the issue.

**Comment by loganmc10** [comment]
This happened to me the other day on Android, but I was unable to reproduce it, just wanted to mention so you guys know that @fzurita isn't crazy lol

It is quite rare though, I've been playing through Paper Mario and using save states and it only happened to me once

**Comment by Dr-Zee** [comment]
I'm on windows 8.1 and suddenly having this exact issue.
It happens, so far, only with Mario 64. 
Retroarch is set to save on exit and auto load state on load.
The game launches perfectly.
I can save a state and load a state in game.
I exit, re-launch, and it crashes when it tries to load the auto save state.
Additionally, the auto save state is 16mb. 

```
RetroArch [INFO] :: Using font rendering backend: freetype.
RetroArch [INFO] :: Initializing rewind buffer with size: 20 MB
RetroArch [WARN] :: Input device ID 5 is unknown to this libretro implementation. Using RETRO_DEVICE_JOYPAD.
RetroArch [INFO] :: Trying to write to playlist file: D:\Games\Retroarch\content_history.lpl
RetroArch [INFO] :: Trying to write to playlist file: D:\Games\Retroarch\content_video_history.lpl
RetroArch [INFO] :: Trying to write to playlist file: D:\Games\Retroarch\content_music_history.lpl
RetroArch [INFO] :: Trying to write to playlist file: D:\Games\Retroarch\content_image_history.lpl
RetroArch [INFO] :: Loading history file: [D:\Games\Retroarch\content_history.lpl].
RetroArch [INFO] :: Loading history file: [D:\Games\Retroarch\content_music_history.lpl].
RetroArch [INFO] :: Loading history file: [D:\Games\Retroarch\content_video_history.lpl].
RetroArch [INFO] :: Loading history file: [D:\Games\Retroarch\content_image_history.lpl].
RetroArch [INFO] :: [GL]: VSync => on
RetroArch [INFO] :: [WGL]: wglSwapInterval(3)
RetroArch [INFO] :: Trying to write to playlist file: D:\Games\Retroarch\content_history.lpl
RetroArch [INFO] :: Environ SYSTEM_DIRECTORY: "D:\Games\Retroarch\system".
RetroArch [INFO] :: Environ SYSTEM_DIRECTORY: "D:\Games\Retroarch\system".
RetroArch [INFO] :: Environ SYSTEM_DIRECTORY: "D:\Games\Retroarch\system".
RetroArch [INFO] :: Loading state: "D:\Games\Retroarch\states\Mupen64Plus OpenGL\Super Mario 64 (English).state.auto".
RetroArch [INFO] :: State size: 16789316 bytes.
RetroArch [INFO] :: Saving state: "RAM".
RetroArch [INFO] :: State size: 16789316 bytes.

<CRASH>
```
If I delete the auto-state, I can launch the game again with no problem.

**Comment by fzurita** [comment]
I'm thinking this game must be mapping TLB memory before attempting to use it. I wonder if TLB mappings are not being saved properly with the new dynarec. I'm thinking that the pause screen expects that memory to be allocated already, and if it's not, it throws that "TLB exception".

Maybe we should implement a speedhack like we do for goldeneye for all TLB games. Any TLB be read always succeeds, and it's mapped right there on the spot.

**Comment by fzurita** [comment]
This is probably fixed.
